src/tool/xmeta/cpp/ast_to_st_listener.cpp

```diff
@@ -169,20 +241,53 @@ listener_error ast_to_st_listener::resolve_enum_val(enum_member& e_member, std::
     return listener_error::passed;
 }
 
+void ast_to_st_listener::exitDelegate_declaration(XlangParser::Delegate_declarationContext* ctx)
+{
+    auto id = ctx->IDENTIFIER();
+    std::string delegate_name{ id->getText() };
+    auto decl_line = id->getSymbol()->getLine();
+
+    std::string symbol = m_reader.get_cur_namespace_body()->get_containing_namespace()->get_fully_qualified_id() + "." + delegate_name;
+    if (m_reader.type_declaration_exists(symbol))
+    {
+        m_reader.write_redeclaration_error(symbol, decl_line);
+        return;
+    }
+
+    std::optional<type_ref> tr = type_ref{ ctx->return_type()->getText() };
+    extract_type(ctx->return_type(), tr);
+    auto dm = std::make_shared<delegate_model>(delegate_name, decl_line, m_reader.get_cur_assembly(), m_reader.get_cur_namespace_body(), std::move(tr));
+
+    auto formal_params = ctx->formal_parameter_list();
+    if (formal_params)
+    {
+        extract_formal_params(formal_params->fixed_parameter(), dm);
+    }
+
+    m_reader.m_cur_namespace_body->add_delegate(dm);
+    m_reader.symbols[symbol] = dm;
```

Comment:
I think a flat table available to all the model is going to be as viable and simpler. A symbol table for each namespace can also get complicated because then you will have to traverse through each namespace to find the model. This way you will have access to the models easily. I already have an idea on how to do the using-directive and it is going to be up to the models to figure out which namespace to append the ref name to since each namespace model will have a list of using directives and aliases.